### PR TITLE
Add Storm API for extended model types (SYN-8187, SYN-4498)

### PR DESCRIPTION
**Pull request by Cisphyx**

### CODE REVIEW

File: synapse/cortex.py [hunk 1]
```diff
@@ -1350,6 +1350,17 @@ def _initCorePerms(self):
              'desc': 'Controls access to deleting specific extended model forms.',
              'ex': 'model.form.del._foo:bar'},
 
+            {'perm': ('model', 'type', 'add'), 'gate': 'cortex',
+             'desc': 'Controls access to adding extended model types.'},
+            {'perm': ('model', 'type', 'add', '<type>'), 'gate': 'cortex',
+             'desc': 'Controls access to adding specific extended model types.',
```
[the remark this answers is not in the document]
Comment (MichaelSquires) [reply synapse/cortex.py:1356]: nm.

File: synapse/cortex.py [hunk 2]
```diff
@@ -3462,6 +3504,68 @@ async def _delForm(self, formname):
...
+            mesg = 'Type info should be a dict.'
+            raise s_exc.BadArg(type=typename, mesg=mesg)
+
+        if not typename.startswith('_'):
+            mesg = 'Extended type must begin with "_".'
+            raise s_exc.BadTypeDef(type=typename, mesg=mesg)
+
+        if self.model.type(typename) is not None:
+            raise s_exc.DupTypeName.init(typename)
+
```
Comment (invisig0th) [inline synapse/cortex.py:3522]: we could clone the type here to make let the base type take a swipe at the opts

File: synapse/lib/stormlib/modelext.py [hunk 3]
```diff
@@ -105,6 +105,21 @@ class LibModelExt(s_stormtypes.Lib):
...
+        {'name': 'addType', 'desc': 'Add an extended type definition to the data model.',
+         'type': {'type': 'function', '_funcname': 'addType',
+                  'args': (
+                      {'name': 'typename', 'type': 'str', 'desc': 'The name of the type to add.'},
+                      {'name': 'basetype', 'type': 'str', 'desc': 'The base type the type is derived from.'},
+                      {'name': 'typeopts', 'type': 'dict', 'desc': 'A Synapse type opts dictionary.'},
+                      {'name': 'typeinfo', 'type': 'dict', 'desc': 'A Synapse type info dictionary.'},
+                  ),
+                  'returns': {'type': 'null'}}},
+        {'name': 'delForm', 'desc': 'Remove an extended type definition from the model.',
```
Comment (invisig0th) [inline synapse/lib/stormlib/modelext.py:117]: `delForm` -> `delType`

**Review by invisig0th — CHANGES_REQUESTED** [review]

File: synapse/cortex.py (hunk 2 above)
Comment (Cisphyx) [reply synapse/cortex.py:3522]: Added a check that the basetype exists and cloned to verify opts.

**Review by invisig0th — APPROVED** [review]

**Pull request merged**